sklearn/ensemble/_hist_gradient_boosting/utils.pyx

```diff
@@ -0,0 +1,151 @@
+        'min_split_gain': 0,
+        'verbosity': 10 if sklearn_params['verbose'] else -10,
+        'boost_from_average': True,
+        'enable_bundle': False,  # also makes feature order consistent
+        'min_data_in_bin': 1,
+        'subsample_for_bin': _BinMapper().subsample,
+    }
+
+    if sklearn_params['loss'] == 'categorical_crossentropy':
+        # LightGBM multiplies hessians by 2 in multiclass loss.
+        lightgbm_params['min_sum_hessian_in_leaf'] *= 2
+        lightgbm_params['learning_rate'] *= 2
+
+    # XGB
+    xgboost_loss_mapping = {
+        'least_squares': 'reg:linear',
+        'binary_crossentropy': 'reg:logistic',
+        'categorical_crossentropy': 'multi:softmax'
+    }
+
+    xgboost_params = {
+        'tree_method': 'hist',
+        'grow_policy': 'lossguide',  # so that we can set max_leaves
+        'objective': xgboost_loss_mapping[sklearn_params['loss']],
+        'learning_rate': sklearn_params['learning_rate'],
+        'n_estimators': sklearn_params['max_iter'],
+        'max_leaves': sklearn_params['max_leaf_nodes'],
+        'max_depth': sklearn_params['max_depth'] or 0,
+        'lambda': sklearn_params['l2_regularization'],
+        'max_bin': sklearn_params['max_bins'],
+        'min_child_weight': 1e-3,
+        'verbosity': 2 if sklearn_params['verbose'] else 0,
+        'silent': sklearn_params['verbose'] == 0,
+        'n_jobs': -1,
+    }
+
+    # Catboost
+    catboost_loss_mapping = {
+        'least_squares': 'RMSE',
+        'binary_crossentropy': 'Logloss',
+        'categorical_crossentropy': 'MultiClass'
+    }
+
+    catboost_params = {
+        'loss_function': catboost_loss_mapping[sklearn_params['loss']],
+        'learning_rate': sklearn_params['learning_rate'],
+        'iterations': sklearn_params['max_iter'],
+        'depth': sklearn_params['max_depth'],
+        'reg_lambda': sklearn_params['l2_regularization'],
+        'max_bin': sklearn_params['max_bins'],
+        'feature_border_type': 'Median',
+        'leaf_estimation_method': 'Newton',
+        'verbose': bool(sklearn_params['verbose']),
+    }
+
+    if lib == 'lightgbm':
+        from lightgbm import LGBMRegressor
+        from lightgbm import LGBMClassifier
+        if is_classifier(estimator):
+            return LGBMClassifier(**lightgbm_params)
+        else:
+            return LGBMRegressor(**lightgbm_params)
+
+    elif lib == 'xgboost':
+        from xgboost import XGBRegressor
+        from xgboost import XGBClassifier
+        if is_classifier(estimator):
+            return XGBClassifier(**xgboost_params)
+        else:
+            return XGBRegressor(**xgboost_params)
+
+    else:
+        from catboost import CatBoostRegressor
+        from catboost import CatBoostClassifier
+        if is_classifier(estimator):
+            return CatBoostClassifier(**catboost_params)
+        else:
+            return CatBoostRegressor(**catboost_params)
+
+
+def sum_parallel(G_H_DTYPE_C [:] array):
```

Comment:
The parallel loop is nogil but the function cannot be nogil since it's called from python.

inlining it wouldn't be useful anyway since it's only called twice per tree growing process.